### PR TITLE
Make IndexSearcher#getSlices final and clarify docs

### PR DESCRIPTION
**Pull request by javanna**
IndexSearcher exposes a public getSlices method, that is used to retriefve the slices that the searcher executes queries against, as well as slices, which is supposed to be overridden to customize the creation of slices.

I believe that getSlices should be final: there is no reason to override the method. Also, it is too easy to confuse the two and end up overriding the wrong one by mistake.

This change was discussed here: https://github.com/apache/lucene/pull/12606#discussion_r1340606929 .

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java [hunk 1]
```diff
@@ -425,11 +425,12 @@ public int count(Query query) throws IOException {
...
 
   /**
-   * Returns the leaf slices used for concurrent searching
+   * Returns the leaf slices used for concurrent searching. Override {@link #slices(List)} to
+   * customize how slices are created.
    *
    * @lucene.experimental
    */
-  public LeafSlice[] getSlices() {
+  public final LeafSlice[] getSlices() {
```
Comment (javanna) [inline lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java:433]: I wonder if we are good backporting this to 9.x given that the method is experimental. I don't see a reason why one would override getSlices on purpose.

File: lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java [hunk 2]
```diff
@@ -115,10 +115,10 @@ public class IndexSearcher {
   protected final List<LeafReaderContext> leafContexts;
 
   /**
-   * used with executor - LeafSlice supplier where each slice holds a set of leafs executed within
+   * Used with executor - LeafSlice supplier where each slice holds a set of leafs executed within
    * one thread. We are caching it instead of creating it eagerly to avoid calling a protected
-   * method from constructor, which is a bad practice. This is {@code null} if no executor is
-   * provided
+   * method from constructor, which is a bad practice. Always non-null, regardless of whether an
+   * executor is provided or not.
```
Comment (javanna) [inline lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java:121]: the behaviour here has recently changed, but the docs were out of date.

**Review by shubhamvishu — APPROVED** [review]
Looks good @javanna!

File: lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java (hunk 1 above)
Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java:433]: ++

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by javanna** [comment]
thanks @s1monw for the review! I will add a changes entry and merge this.

**Pull request merged**